File: unity/Assets/Scripts/Multiplayer/Rooms/Realtime/Realtime.cs
using System;
using System.Net.WebSockets;
using System.Text;
using System.Threading;
using System.Threading.Tasks;
using UnityEngine;
using Michitai.Multiplayer;
using System.Net.Http;

namespace Michitai.Multiplayer.Rooms.Realtime
{
    /// <summary>
    /// Manages WebSocket connections for realtime communication in game rooms in Unity.
    /// Handles connection, message sending/receiving, and automatic heartbeat.
    /// Uses Unity's JsonUtility for serialization.
    /// </summary>
    public class Realtime
    {
        private ClientWebSocket _websocket;
        private string _token;
        private CancellationTokenSource _cancellationTokenSource;
        private CancellationTokenSource _heartbeatCancellationTokenSource;
        private readonly string _realtimeWebSocketUrl;
        private bool _isConnecting = false;

        /// <summary>
        /// Event raised when a message is received from the WebSocket.
        /// </summary>
        public event Action<string, string, SenderInfo> OnReceive;

        /// <summary>
        /// Event raised when the WebSocket connection is established.
        /// </summary>
        public event Action OnConnected;

        /// <summary>
        /// Initializes a new Realtime instance.
        /// </summary>
        /// <param name="realtimeWebSocketUrl">The WebSocket server URL (default: wss://realtime.michitai.com).</param>
        public Realtime(string realtimeWebSocketUrl = "wss://realtime.michitai.com")
        {
            _realtimeWebSocketUrl = realtimeWebSocketUrl;
        }

        /// <summary>
        /// Retrieves a realtime authentication token for WebSocket connections.
        /// </summary>
        /// <param name="client">The API client instance.</param>
        /// <param name="playerToken">The player's private authentication token.</param>
        /// <returns>Response containing the realtime token.</returns>
        public static async Task<TokenResponse> GetTokenAsync(Client client, string playerToken)
        {
            var url = client.Url(Endpoints.RealtimeToken, $"&player_token={playerToken}");
            return await client.Send<TokenResponse>(HttpMethod.Post, url, null);
        }

        /// <summary>
        /// Connects to the realtime WebSocket server using the provided token.
        /// </summary>
        /// <param name="realtimeToken">The realtime authentication token.</param>
        /// <returns>True if connection succeeded, false otherwise.</returns>
        public async Task<bool> ConnectAsync(string realtimeToken)
        {
            if (_isConnecting || (_websocket?.State == WebSocketState.Open))
                return false;

            _isConnecting = true;
            try
            {
                _cancellationTokenSource = new CancellationTokenSource();
                _token = realtimeToken;

                // Wake up the server before connecting
                using (var httpClient = new HttpClient())
                {
                    await httpClient.GetAsync("https://realtime.michitai.com/");
                }

                var uri = new Uri($"{_realtimeWebSocketUrl}?token={_token}&client=unity");

                _websocket = new ClientWebSocket();

                await _websocket.ConnectAsync(uri, _cancellationTokenSource.Token);

                _ = Task.Run(ListenForMessagesAsync, _cancellationTokenSource.Token);
                _ = Task.Run(StartHeartbeatAsync, _cancellationTokenSource.Token);

                OnConnected?.Invoke();

                return true;
            }
            catch (Exception ex)
            {
                Debug.LogError($"Connection failed: {ex.Message}");
                await DisconnectAsync();
                return false;
            }
            finally
            {
                _isConnecting = false;
            }
        }

        /// <summary>
        /// Sends a message to the specified players via WebSocket.
        /// Uses Unity's JsonUtility for serialization.
        /// </summary>
        /// <typeparam name="T">The type of data to send.</typeparam>
        /// <param name="target">The target players (All, Host, Others, Specific).</param>
        /// <param name="command">The command/type of the message.</param>
        /// <param name="data">Optional data payload to send.</param>
        /// <param name="targetIds">Specific player IDs if target is Specific.</param>
        public async Task SendAsync<T>(ERoomTargetPlayer target, string command, T data = null, int[] targetIds = null) where T : class, new()
        {
            if (_websocket?.State != WebSocketState.Open) return;

            var message = new SendMessage
            {
                type = "send",
                command = command,
                data_json = data != null ? JsonUtility.ToJson(data) : null,
                target_ids = targetIds ?? new int[0],
                target = target.ToString().ToLower()
            };

            var json = JsonUtility.ToJson(message);
            var buffer = Encoding.UTF8.GetBytes(json);

            await _websocket.SendAsync(
                new ArraySegment<byte>(buffer),
                WebSocketMessageType.Text,
                true,
                _cancellationTokenSource?.Token ?? CancellationToken.None);
        }

        /// <summary>
        /// Disconnects from the WebSocket server and cleans up resources.
        /// </summary>
        public async Task DisconnectAsync()
        {
            try
            {
                _heartbeatCancellationTokenSource?.Cancel();
                _cancellationTokenSource?.Cancel();

                if (_websocket?.State == WebSocketState.Open)
                {
                    await _websocket.CloseAsync(
                        WebSocketCloseStatus.NormalClosure,
                        "Client disconnecting",
                        CancellationToken.None);
                }
            }
            catch (Exception ex)
            {
                Debug.LogError($"Disconnect error: {ex.Message}");
            }
        }

        private async Task ListenForMessagesAsync()
        {
            var buffer = new byte[4096];

            while (_websocket?.State == WebSocketState.Open &&
                !(_cancellationTokenSource?.Token.IsCancellationRequested ?? true))
            {
                try
                {
                    var result = await _websocket.ReceiveAsync(
                        new ArraySegment<byte>(buffer),
                        _cancellationTokenSource?.Token ?? CancellationToken.None);

                    if (result.MessageType == WebSocketMessageType.Text && result.Count > 0)
                    {
                        var messageStr = Encoding.UTF8.GetString(buffer, 0, result.Count);
                        var realtimeMessage = JsonUtility.FromJson<RealtimeMessage>(messageStr);

                        if (realtimeMessage?.type == "receive" && OnReceive != null)
                        {
                            OnReceive.Invoke(realtimeMessage.command ?? string.Empty,
                                realtimeMessage.data_json ?? string.Empty,
                                realtimeMessage.sender);
                        }
                    }
                }
                catch (Exception ex) when (!(ex is OperationCanceledException))
                {
                    Debug.LogError($"Message receive error: {ex.Message}");
                    if (ex is WebSocketException) break;
                }
            }
        }

        private async Task StartHeartbeatAsync()
        {
            _heartbeatCancellationTokenSource = new CancellationTokenSource();
            
            while (!_heartbeatCancellationTokenSource.Token.IsCancellationRequested && 
                   _websocket?.State == WebSocketState.Open)
            {
                try
                {
                    await Task.Delay(TimeSpan.FromSeconds(20), _heartbeatCancellationTokenSource.Token);
                    
                    if (!_heartbeatCancellationTokenSource.Token.IsCancellationRequested && 
                        _websocket?.State == WebSocketState.Open)
                    {
                        await SendHeartbeatAsync();
                    }
                }
                catch (OperationCanceledException)
                {
                    break;
                }
                catch (Exception ex)
                {
                    Debug.LogError($"Heartbeat error: {ex.Message}");
                }
            }
        }

        private async Task SendHeartbeatAsync()
        {
            try
            {
                var message = new SendMessage
                {
                    type = "heartbeat"
                };

                var json = JsonUtility.ToJson(message);
                var buffer = Encoding.UTF8.GetBytes(json);

                await _websocket.SendAsync(
                    new ArraySegment<byte>(buffer),
                    WebSocketMessageType.Text,
                    true,
                    _heartbeatCancellationTokenSource?.Token ?? CancellationToken.None);
            }
            catch (Exception ex)
            {
                Debug.LogError($"Send heartbeat error: {ex.Message}");
            }
        }

        private void OnDestroy()
        {
            _ = DisconnectAsync();
        }
    }
}

File: java/src/main/java/com/michitai/multiplayer/rooms/realtime/Realtime.java
package com.michitai.multiplayer.rooms.realtime;

import com.fasterxml.jackson.databind.ObjectMapper;
import com.michitai.multiplayer.Client;

import java.io.IOException;
import java.net.URI;
import java.net.http.HttpClient;
import java.net.http.HttpRequest;
import java.net.http.HttpResponse;
import java.net.http.WebSocket;
import java.nio.ByteBuffer;
import java.util.concurrent.CompletableFuture;
import java.util.concurrent.ExecutionException;
import java.util.concurrent.TimeUnit;
import java.util.concurrent.atomic.AtomicBoolean;

/**
 * Manages WebSocket connections for realtime communication in game rooms.
 * Handles connection, message sending/receiving, and automatic heartbeat.
 */
public class Realtime {
    private WebSocket webSocket;
    private String token;
    private CompletableFuture<Void> messageListenerFuture;
    private CompletableFuture<Void> heartbeatFuture;
    private final String realtimeWebSocketUrl;
    private final ObjectMapper objectMapper;
    private final AtomicBoolean isConnected = new AtomicBoolean(false);

    /**
     * Event raised when a message is received from the WebSocket.
     */
    public interface MessageListener {
        void onMessage(String command, Object data, SenderInfo sender);
    }

    /**
     * Event raised when the WebSocket connection is established.
     */
    public interface ConnectionListener {
        void onConnected();
    }

    private MessageListener messageListener;
    private ConnectionListener connectionListener;

    /**
     * Initializes a new Realtime instance.
     *
     * @param realtimeWebSocketUrl The WebSocket server URL (default: wss://realtime.michitai.com).
     */
    public Realtime(String realtimeWebSocketUrl) {
        this.realtimeWebSocketUrl = realtimeWebSocketUrl != null ? realtimeWebSocketUrl : "wss://realtime.michitai.com";
        this.objectMapper = Client.JSON_MAPPER;
    }

    public Realtime() {
        this("wss://realtime.michitai.com");
    }

    /**
     * Sets the message listener for received messages.
     */
    public void setMessageListener(MessageListener listener) {
        this.messageListener = listener;
    }

    /**
     * Sets the connection listener for connection events.
     */
    public void setConnectionListener(ConnectionListener listener) {
        this.connectionListener = listener;
    }

    /**
     * Retrieves a realtime authentication token for WebSocket connections.
     *
     * @param client The API client instance.
     * @param playerToken The player's private authentication token.
     * @return Response containing the realtime token.
     * @throws IOException if the request fails.
     */
    public static TokenResponse getToken(Client client, String playerToken) throws IOException {
        String url = client.url(Endpoints.REALTIME_TOKEN, "&player_token=" + playerToken);
        return client.post(url, null, TokenResponse.class);
    }

    /**
     * Connects to the realtime WebSocket server using the provided token.
     *
     * @param realtimeToken The realtime authentication token.
     * @return True if connection succeeded, false otherwise.
     */
    public boolean connect(String realtimeToken) {
        try {
            this.token = realtimeToken;
            isConnected.set(false);

            // Wake up the server before connecting
            HttpClient httpClient = HttpClient.newHttpClient();
            HttpRequest wakeRequest = HttpRequest.newBuilder()
                .uri(URI.create("https://realtime.michitai.com/"))
                .GET()
                .build();
            httpClient.send(wakeRequest, HttpResponse.BodyHandlers.ofString());

            URI uri = URI.create(realtimeWebSocketUrl + "?token=" + token + "&client=json");
            
            WebSocket.Listener listener = new WebSocket.Listener() {
                private StringBuilder messageBuilder = new StringBuilder();

                @Override
                public CompletableFuture<Void> onText(WebSocket webSocket, CharSequence data, boolean last) {
                    messageBuilder.append(data);
                    if (last) {
                        String message = messageBuilder.toString();
                        messageBuilder.setLength(0);
                        handleMessage(message);
                    }
                    return CompletableFuture.completedFuture(null);
                }

                @Override
                public void onOpen(WebSocket webSocket) {
                    isConnected.set(true);
                    if (connectionListener != null) {
                        connectionListener.onConnected();
                    }
                    webSocket.request(1);
                }
            };

            webSocket = HttpClient.newHttpClient()
                .newWebSocketBuilder()
                .buildAsync(uri, listener)
                .join();

            // Start message listener
            messageListenerFuture = CompletableFuture.runAsync(() -> {
                while (isConnected.get()) {
                    try {
                        TimeUnit.MILLISECONDS.sleep(100);
                    } catch (InterruptedException e) {
                        break;
                    }
                }
            });

            // Start heartbeat
            heartbeatFuture = CompletableFuture.runAsync(this::heartbeatLoop);

            return true;
        } catch (Exception ex) {
            System.err.println("Connection failed: " + ex.getMessage());
            disconnect();
            return false;
        }
    }

    /**
     * Sends a message to the specified players via WebSocket.
     *
     * @param target The target players (ALL, HOST, OTHERS, SPECIFIC).
     * @param command The command/type of the message.
     * @param data Optional data payload to send.
     * @param targetIds Specific player IDs if target is SPECIFIC.
     */
    public void send(ERoomTargetPlayer target, String command, Object data, int[] targetIds) {
        if (!isConnected.get() || webSocket == null) return;

        try {
            SendMessage message = new SendMessage();
            message.setType("send");
            message.setCommand(command);
            message.setData(data);
            message.setTargetIds(targetIds != null ? targetIds : new int[0]);
            message.setTarget(target.name().toLowerCase());

            String json = objectMapper.writeValueAsString(message);
            webSocket.sendText(json, true);
        } catch (Exception ex) {
            System.err.println("Send error: " + ex.getMessage());
        }
    }

    /**
     * Disconnects from the WebSocket server and cleans up resources.
     */
    public void disconnect() {
        try {
            if (heartbeatFuture != null) {
                heartbeatFuture.cancel(true);
            }
            if (messageListenerFuture != null) {
                messageListenerFuture.cancel(true);
            }

            if (webSocket != null) {
                webSocket.sendClose(WebSocket.NORMAL_CLOSURE, "Client disconnecting").join();
            }
        } catch (Exception ex) {
            System.err.println("Disconnect error: " + ex.getMessage());
        } finally {
            isConnected.set(false);
        }
    }

    private void handleMessage(String message) {
        try {
            RealtimeMessage realtimeMessage = objectMapper.readValue(message, RealtimeMessage.class);
            
            if ("receive".equals(realtimeMessage.getType()) && messageListener != null) {
                messageListener.onMessage(
                    realtimeMessage.getCommand() != null ? realtimeMessage.getCommand() : "",
                    realtimeMessage.getData() != null ? realtimeMessage.getData() : new Object(),
                    realtimeMessage.getSender()
                );
            }
        } catch (Exception ex) {
            System.err.println("Message handling error: " + ex.getMessage());
        }
    }

    private void heartbeatLoop() {
        while (isConnected.get()) {
            try {
                TimeUnit.SECONDS.sleep(20);
                
                if (isConnected.get() && webSocket != null) {
                    sendHeartbeat();
                }
            } catch (InterruptedException e) {
                break;
            } catch (Exception ex) {
                System.err.println("Heartbeat error: " + ex.getMessage());
            }
        }
    }

    private void sendHeartbeat() {
        try {
            HeartbeatMessage message = new HeartbeatMessage();
            message.setType("heartbeat");
            
            String json = objectMapper.writeValueAsString(message);
            webSocket.sendText(json, true);
        } catch (Exception ex) {
            System.err.println("Send heartbeat error: " + ex.getMessage());
        }
    }

    /**
     * Inner class for send messages.
     */
    private static class SendMessage {
        private String type;
        private String command;
        private Object data;
        private int[] targetIds;
        private String target;

        public String getType() { return type; }
        public void setType(String type) { this.type = type; }
        public String getCommand() { return command; }
        public void setCommand(String command) { this.command = command; }
        public Object getData() { return data; }
        public void setData(Object data) { this.data = data; }
        public int[] getTargetIds() { return targetIds; }
        public void setTargetIds(int[] targetIds) { this.targetIds = targetIds; }
        public String getTarget() { return target; }
        public void setTarget(String target) { this.target = target; }
    }

    /**
     * Inner class for heartbeat messages.
     */
    private static class HeartbeatMessage {
        private String type;

        public String getType() { return type; }
        public void setType(String type) { this.type = type; }
    }

    public boolean isConnected() {
        return isConnected.get();
    }
}
